Can't use progress_hook with DataLakeFileClient.upload_data

- **Package Name**: azure-storage-file-datalake
- **Package Version**: 12.18.0
- **Operating System**: Windows 11
- **Python Version**: 3.11.7

**Describe the bug**
azure.storage.filedatalake.aio.DataLakeFileClient doesn't support progress_hook parameter, when using it, there is error like:

```
    def request(
        self, method: str, url: StrOrURL, **kwargs: Any
    ) -> "_RequestContextManager":
        """Perform HTTP request."""
>       return _RequestContextManager(self._request(method, url, **kwargs))
E       TypeError: ClientSession._request() got an unexpected keyword argument 'progress_hook'
```

But the upload_data_chunks supports progress_hook.

**To Reproduce**
```
    from azure.core import MatchConditions

    async def passthrough(total, received):
        print((total, received))
    async with AsyncDataLakeFileClient(
       account_url,
        "filesystem",
        "/local/users/testdata/file.txt",
        creds,
    ) as fc:
        await fc.upload_data(
            b"1,2,3",
            overwrite=True,
            match_condition=MatchConditions.IfMissing,
            max_concurrency=12,
            progress_hook=passthrough,
        )
```

**Expected behavior**
progress_hook to work without error.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

--- a/sdk/storage/azure-storage-file-datalake/assets.json
+++ b/sdk/storage/azure-storage-file-datalake/assets.json
@@ -2,5 +2,5 @@
   "AssetsRepo": "Azure/azure-sdk-assets",
   "AssetsRepoPrefixPath": "python",
   "TagPrefix": "python/storage/azure-storage-file-datalake",
-  "Tag": "python/storage/azure-storage-file-datalake_caec744857"
+  "Tag": "python/storage/azure-storage-file-datalake_a7ee812abc"
 }

--- a/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_file_client.py
+++ b/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_file_client.py
@@ -457,6 +457,11 @@ class DataLakeFileClient(PathClient):
             Defaults to 100*1024*1024, or 100MB.
         :keyword str encryption_context:
             Specifies the encryption context to set on the file.
+        :keyword progress_hook:
+            A callback to track the progress of a long-running upload. The signature is
+            function(current: int, total: int) where current is the number of bytes transferred
+            so far, and total is the total size of the download.
+        :paramtype progress_hook: Callable[[int, int], None]
         :returns: A dictionary of response headers.
         :rtype: Dict[str, Any]
         """
@@ -683,6 +688,11 @@ class DataLakeFileClient(PathClient):
             Maximum number of parallel connections to use when transferring the file in chunks.
             This option does not affect the underlying connection pool, and may
             require a separate configuration of the connection pool.
+        :keyword progress_hook:
+            A callback to track the progress of a long-running download. The signature is
+            function(current: int, total: int) where current is the number of bytes transferred
+            so far, and total is the total size of the download.
+        :paramtype progress_hook: Callable[[int, int], None]
         :keyword int timeout:
             Sets the server-side timeout for the operation in seconds. For more details see
             https://learn.microsoft.com/rest/api/storageservices/setting-timeouts-for-blob-service-operations.

--- a/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_upload_helper.py
+++ b/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_upload_helper.py
@@ -53,6 +53,7 @@ def upload_datalake_file(
         modified_access_conditions = kwargs.pop('modified_access_conditions', None)
         chunk_size = kwargs.pop('chunk_size', 100 * 1024 * 1024)
         encryption_context = kwargs.pop('encryption_context', None)
+        progress_hook = kwargs.pop('progress_hook', None)
 
         if not overwrite:
             # if customers didn't specify access conditions, they cannot flush data to existing file
@@ -95,6 +96,7 @@ def upload_datalake_file(
                 stream=stream,
                 max_concurrency=max_concurrency,
                 validate_content=validate_content,
+                progress_hook=progress_hook,
                 **kwargs
             )
         else:
@@ -106,6 +108,7 @@ def upload_datalake_file(
                 max_concurrency=max_concurrency,
                 stream=stream,
                 validate_content=validate_content,
+                progress_hook=progress_hook,
                 **kwargs
             )
 

--- a/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_file_client_async.py
+++ b/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_file_client_async.py
@@ -473,7 +473,12 @@ class DataLakeFileClient(PathClient):
             Defaults to 100*1024*1024, or 100MB.
         :keyword str encryption_context:
             Specifies the encryption context to set on the file.
-        :return: A dictionary of response headers.
+        :keyword progress_hook:
+            A callback to track the progress of a long-running upload. The signature is
+            function(current: int, total: int) where current is the number of bytes transferred
+            so far, and total is the total size of the download.
+        :paramtype progress_hook: Callable[[int, Optional[int]], Awaitable[None]]
+        :returns: A dictionary of response headers.
         :rtype: Dict[str, Any]
         """
         options = _upload_options(
@@ -699,6 +704,11 @@ class DataLakeFileClient(PathClient):
             Maximum number of parallel connections to use when transferring the file in chunks.
             This option does not affect the underlying connection pool, and may
             require a separate configuration of the connection pool.
+        :keyword progress_hook:
+            A callback to track the progress of a long-running download. The signature is
+            function(current: int, total: int) where current is the number of bytes transferred
+            so far, and total is the total size of the download.
+        :paramtype progress_hook: Callable[[int, Optional[int]], Awaitable[None]]
         :keyword int timeout:
             Sets the server-side timeout for the operation in seconds. For more details see
             https://learn.microsoft.com/rest/api/storageservices/setting-timeouts-for-blob-service-operations.

--- a/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_upload_helper.py
+++ b/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_upload_helper.py
@@ -53,6 +53,7 @@ async def upload_datalake_file(
         modified_access_conditions = kwargs.pop('modified_access_conditions', None)
         chunk_size = kwargs.pop('chunk_size', 100 * 1024 * 1024)
         encryption_context = kwargs.pop('encryption_context', None)
+        progress_hook = kwargs.pop('progress_hook', None)
 
         if not overwrite:
             # if customers didn't specify access conditions, they cannot flush data to existing file
@@ -95,6 +96,7 @@ async def upload_datalake_file(
                 stream=stream,
                 max_concurrency=max_concurrency,
                 validate_content=validate_content,
+                progress_hook=progress_hook,
                 **kwargs
             )
         else:
@@ -106,6 +108,7 @@ async def upload_datalake_file(
                 max_concurrency=max_concurrency,
                 stream=stream,
                 validate_content=validate_content,
+                progress_hook=progress_hook,
                 **kwargs
             )
 

--- a/sdk/storage/azure-storage-file-datalake/tests/test_file.py
+++ b/sdk/storage/azure-storage-file-datalake/tests/test_file.py
@@ -37,7 +37,7 @@ from azure.storage.filedatalake import (
 from devtools_testutils import recorded_by_proxy
 from devtools_testutils.storage import StorageRecordedTestCase
 from settings.testcase import DataLakePreparer
-from test_helpers import MockStorageTransport
+from test_helpers import MockStorageTransport, ProgressTracker
 
 # ------------------------------------------------------------------------------
 
@@ -1691,6 +1691,33 @@ class TestFile(StorageRecordedTestCase):
 
         file_data = file_client.download_file(validate_content=True).read()
         assert file_data == b"Hello World!"  # data is fixed by mock transport
+
+    @DataLakePreparer()
+    @recorded_by_proxy
+    def test_progress_hook_upload_data(self, **kwargs):
+        datalake_storage_account_name = kwargs.pop("datalake_storage_account_name")
+        datalake_storage_account_key = kwargs.pop("datalake_storage_account_key")
+
+        # Arrange
+        self._setUp(datalake_storage_account_name, datalake_storage_account_key)
+        file_client = self._create_file_and_return_client(
+            directory=self._get_directory_reference(),
+            file=self._get_file_reference()
+        )
+        data = self.get_random_bytes(8 * 1024)
+        progress = ProgressTracker(len(data), 1024)
+
+        # Act
+        file_client.upload_data(
+            data,
+            overwrite=True,
+            progress_hook=progress.assert_progress,
+            max_concurrency=1,
+            chunk_size=1024
+        )
+
+        # Assert
+        progress.assert_complete()
 
 # ------------------------------------------------------------------------------
 if __name__ == '__main__':

--- a/sdk/storage/azure-storage-file-datalake/tests/test_file_async.py
+++ b/sdk/storage/azure-storage-file-datalake/tests/test_file_async.py
@@ -35,7 +35,7 @@ from azure.storage.filedatalake.aio import DataLakeDirectoryClient, DataLakeFile
 from devtools_testutils.aio import recorded_by_proxy_async
 from devtools_testutils.storage.aio import AsyncStorageRecordedTestCase
 from settings.testcase import DataLakePreparer
-from test_helpers_async import AsyncStream, MockStorageTransport
+from test_helpers_async import AsyncStream, MockStorageTransport, ProgressTracker
 # ------------------------------------------------------------------------------
 
 TEST_DIRECTORY_PREFIX = 'directory'
@@ -1593,6 +1593,33 @@ class TestFileAsync(AsyncStorageRecordedTestCase):
 
         file_data = await (await file_client.download_file(validate_content=True)).read()
         assert file_data == b"Hello Async World!"  # data is fixed by mock transport
+
+    @DataLakePreparer()
+    @recorded_by_proxy_async
+    async def test_progress_hook_upload_data(self, **kwargs):
+        datalake_storage_account_name = kwargs.pop("datalake_storage_account_name")
+        datalake_storage_account_key = kwargs.pop("datalake_storage_account_key")
+
+        # Arrange
+        await self._setUp(datalake_storage_account_name, datalake_storage_account_key)
+        file_client = await self._create_file_and_return_client(
+            directory=self._get_directory_reference(),
+            file=self._get_file_reference()
+        )
+        data = self.get_random_bytes(8 * 1024)
+        progress = ProgressTracker(len(data), 1024)
+
+        # Act
+        await file_client.upload_data(
+            data,
+            overwrite=True,
+            progress_hook=progress.assert_progress,
+            max_concurrency=1,
+            chunk_size=1024
+        )
+
+        # Assert
+        progress.assert_complete()
 
 # ------------------------------------------------------------------------------
 if __name__ == '__main__':

--- a/sdk/storage/azure-storage-file-datalake/tests/test_helpers.py
+++ b/sdk/storage/azure-storage-file-datalake/tests/test_helpers.py
@@ -4,7 +4,7 @@
 # license information.
 # --------------------------------------------------------------------------
 
-from typing import Any, Dict
+from typing import Any, Dict, Optional
 from typing_extensions import Self
 from urllib.parse import urlparse
 
@@ -12,6 +12,24 @@ from azure.core.pipeline.transport import HttpTransport, RequestsTransportRespon
 from azure.core.rest import HttpRequest
 from requests import Response
 from urllib3 import HTTPResponse
+
+
+class ProgressTracker:
+    def __init__(self, total: int, step: int):
+        self.total = total
+        self.step = step
+        self.current = 0
+
+    def assert_progress(self, current: int, total: Optional[int]):
+        if self.current != self.total:
+            self.current += self.step
+
+        if total:
+            assert self.total == total
+        assert self.current == current
+
+    def assert_complete(self):
+        assert self.total == self.current
 
 
 class MockHttpClientResponse(Response):

--- a/sdk/storage/azure-storage-file-datalake/tests/test_helpers_async.py
+++ b/sdk/storage/azure-storage-file-datalake/tests/test_helpers_async.py
@@ -3,12 +3,30 @@
 # Licensed under the MIT License. See License.txt in the project root for
 # license information.
 # --------------------------------------------------------------------------
-from typing import Any, Dict
+from typing import Any, Dict, Optional
 from urllib.parse import urlparse
 
 from azure.core.pipeline.transport import AioHttpTransportResponse, AsyncHttpTransport
 from azure.core.rest import HttpRequest
 from aiohttp import ClientResponse
+
+
+class ProgressTracker:
+    def __init__(self, total: int, step: int):
+        self.total = total
+        self.step = step
+        self.current = 0
+
+    async def assert_progress(self, current: int, total: Optional[int]):
+        if self.current != self.total:
+            self.current += self.step
+
+        if total:
+            assert self.total == total
+        assert self.current == current
+
+    def assert_complete(self):
+        assert self.total == self.current
 
 
 class AsyncStream:

